flowmachine/flowmachine/core/random.py

```diff
@@ -3,293 +3,497 @@
+
+    # Overwrite to call on parent instead
+    @property
+    def column_names(self) -> List[str]:
+        return self.query.column_names
+
+
+class RandomSystemRows(RandomBase):
+    """
+    Gets a random sample from the result of a query, using a PostgreSQL TABLESAMPLE
+    clause with the 'system_rows' method.
+    This method performs block-level sampling by randomly sampling
+    each physical storage page of the underlying relation. This
+    sampling method is guaranteed to provide a sample of the specified
+    size.
+
+    Parameters
+    ----------
+    query : str
+        A query specifying a table from which a random sample will be drawn.
+    size : int, optional
+        The number of rows to be selected from the table.
+        Exactly one of the 'size' or 'fraction' arguments must be provided.
+    fraction : float, optional
+        The fraction of rows to be selected from the table.
+        Exactly one of the 'size' or 'fraction' arguments must be provided.
+    estimate_count : bool, default True
+        Whether to estimate the number of rows in the table using
+        information contained in the `pg_class` or whether to perform an
+        actual count in the number of rows.
+    
+    See Also
+    --------
+    flowmachine.core.random.random_factory
+    
+    Notes
+    -----
+    The 'system_rows' sampling method does not support parent tables which have
+    child inheritance.
+    The 'system_rows' sampling method does not support supplying a seed for
+    reproducible samples, so random samples cannot be stored.
+    """
+
+    def __init__(self, query, *, size=None, fraction=None, estimate_count=True):
+        # Raise a value error if the query is a table, and has children, as the
+        # method relies on it not having children.
+        if isinstance(query, Table) and query.has_children():
             raise ValueError(
                 "It is not possible to use the 'system_rows' method in tables with inheritance "
                 + "as it selects a random sample for each child table and not for the set as a whole."
             )
 
+        super().__init__(
+            query=query, size=size, fraction=fraction, estimate_count=estimate_count
+        )
+
     def _make_query(self):
+        # TABLESAMPLE only works on tables, so silently store this query
+        self.query.store().result()
+
+        columns = ", ".join(
+            [
+                "{}.{}".format(self.query.fully_qualified_table_name, c)
+                for c in self.query.column_names
+            ]
+        )
+
+        if self.size is None:
+            rowcount = self._count_table_rows()
+            size = int(self.fraction * float(rowcount))
+        else:
+            size = self.size
+
+        sampled_query = f"""
+        SELECT {columns} FROM {self.query.fully_qualified_table_name} TABLESAMPLE SYSTEM_ROWS({size})
+        """
+
+        return sampled_query
+
+
+class SeededRandom(RandomBase, metaclass=ABCMeta):
```

Comment:
Let's change this to `SeedableRandom`, because there's no guarantee it is _actually_ seeded.